### ISSUE TITLE
x² shouldn't emit `cast_sign_loss`

### ISSUE THREAD

**Issue by ShE3py** [rust-lang/rust-clippy#11642]
### Summary

Hi,

Clippy shoudn't lint about casting x<sup>2n</sup> as an unsigned integer since the result cannot be negative. The lint should still be shown for x<sup>3</sup>.

### Lint Name

cast_sign_loss

### Reproducer

I tried this code:

```rust
#[forbid(clippy::cast_sign_loss)]
pub fn square(x: i16) -> u32 {
    let x = x as i32;
    
    (x * x) as u32
}
```

I saw this happen:

```
error: casting `i32` to `u32` may lose the sign of the value
 --> src/lib.rs:6:5
  |
6 |     (x * x) as u32
  |     ^^^^^^^^^^^^^^
  |
  = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#cast_sign_loss
```

I expected to see this happen:

No output, as x² &GreaterSlantEqual; 0.

### Version

```text
rustc 1.73.0 (cc66ad468 2023-10-03)
binary: rustc
commit-hash: cc66ad468955717ab92600c770da8c1601a4ff33
commit-date: 2023-10-03
host: x86_64-unknown-linux-gnu
release: 1.73.0
LLVM version: 17.0.2
```
